smithy-trait-codegen/codegen-core/src/main/java/software/amazon/smithy/traitcodegen/generators/common/ToNodeGenerator.java

```diff
@@ -0,0 +1,248 @@
+            writer.write("return values.entrySet().stream()")
+                    .indent()
+                    .write(".map(entry -> new AbstractMap.SimpleImmutableEntry<>(")
+                    .indent()
+                    .write(keyMapper + ", " + valueMapper + "))", "entry.getKey()", "entry.getValue()")
+                    .dedent()
+                    .write(".collect(ObjectNode.collect(Map.Entry::getKey, Map.Entry::getValue))")
+                    .write(".toBuilder().sourceLocation(getSourceLocation()).build();")
+                    .dedent();
+            return null;
+        }
+
+        @Override
+        public Void stringShape(StringShape shape) {
+            writer.addImport(StringNode.class);
+            writer.write("return new StringNode(value.toString(), getSourceLocation());");
+            return null;
+        }
+
+        @Override
+        public Void structureShape(StructureShape shape) {
+            if (!shape.members().isEmpty()) {
+                writer.addImport(Node.class);
+                writer.write("return Node.objectNodeBuilder()").indent();
+                if (shape.hasTrait(TraitDefinition.class)) {
+                    writer.write(".sourceLocation(getSourceLocation())");
+                }
+
+                // Generate all members
+                // TODO: This is all quite clunky. Fix
+                for (MemberShape member : shape.members()) {
+                    Symbol memberSymbol = symbolProvider.toSymbol(member);
+                    memberSymbol.getProperty(SymbolProperties.NODE_MAPPING_IMPORTS, Symbol.class)
+                            .ifPresent(writer::addImport);
+                    if (member.isRequired()) {
+                        writer.write(".withMember($S, "
+                                        + memberSymbol.expectProperty(SymbolProperties.TO_NODE_MAPPER, String.class)
+                                        + ")",
+                                symbolProvider.toMemberName(member), symbolProvider.toMemberName(member));
+                    } else if (model.expectShape(member.getTarget()).isListShape()) {
+                        writer.addImport(ArrayNode.class);
+                        Symbol listTargetSymbol = symbolProvider.toSymbol(
+                                model.expectShape(member.getTarget()).asListShape()
+                                        .orElseThrow(RuntimeException::new).getMember());
+                        writer.write(".withMember($S, get$L().stream().map(s -> "
+                                        + listTargetSymbol.expectProperty(SymbolProperties.TO_NODE_MAPPER, String.class)
+                                        + ").collect(ArrayNode.collect()))",
+                                symbolProvider.toMemberName(member),
+                                StringUtils.capitalize(symbolProvider.toMemberName(member)), "s");
+                    } else if (model.expectShape(member.getTarget()).isMapShape()) {
```

Comment:
I think you are correct about the conditional ordering. I will update to clean up. 

Yes, this currently assumes that rather than getting an `Optional<Collection<?>>` you would get an empty collection. Is that not how we would want these to behave?

Reply:
I mean, technically those are different things. I'm not suuuuper concerned about it though